client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -548,19 +548,27 @@ Response get() throws IOException {
             } catch (InterruptedException e) {
                 throw new RuntimeException("thread waiting for the response was interrupted", e);
             }
-            if (response != null) {
-                assert exception == null;
-                return response;
-            }
-            assert exception != null;
-            //try and leave the exception untouched as much as possible but we don't want to just add throws Exception clause everywhere
-            if (exception instanceof IOException) {
-                throw (IOException) exception;
+
+            if (exception != null) {
+                if (response != null) {
+                    AssertionError e = new AssertionError("response and exception are set at the next time");
```

Comment:
now you can just throw an `IllegaStateException("response and exception are set at the next time", exception)`